core/controllers/ingest.py

```diff
@@ -59,63 +63,194 @@ def ingest(body: dict, excel_file: FileStorage) -> tuple[dict, int]:
+        current_app.logger.error(
+            f"Uncaught ingest exception: {type(uncaught_exception).__name__}: {str(uncaught_exception)},"
+            f" - failure_id={str(failure_uuid)}",
+            exc_info=True,
+        )
+        return build_internal_error_response(
+            detail=f"Uncaught ingest exception: {type(uncaught_exception).__name__}: {str(uncaught_exception)}",
+            failure_uuid=failure_uuid,
+        )
 
     if do_load:
-        load_data(transformed_data, excel_file, reporting_round)
-
-    success_payload = {
-        "detail": f"Spreadsheet successfully validated{' and ingested' if do_load else ' but NOT ingested'}",
-        "status": 200,
-        "title": "success",
-        "metadata": get_metadata(transformed_data, reporting_round),
-        "loaded": do_load,
-    }
+        load_data(transformed_data, excel_file)
+
+    programme_metadata = get_metadata(transformed_data)
+    return build_success_response(programme_metadata=programme_metadata, do_load=do_load)
 
-    return success_payload, 200
 
+def parse_body(body: dict) -> tuple[str, int, dict | None, bool]:
+    """Parses the request body.
 
-def process_initial_validation_errors(error_messages: list[str]) -> tuple[dict, int]:
+    :param body: request body
+    :return: parsed values
+    """
+    fund = body.get("fund_name")
+    reporting_round = body.get("reporting_round")
+    auth = parse_auth(body)
+    do_load = body.get("do_load", True)  # defaults to True, if False then do not load to database
+    return fund, reporting_round, auth, do_load
+
+
+def extract_process_validate_tables(
+    workbook_data: dict[str, pd.DataFrame], tables_config: dict[str, dict]
+) -> dict[str, Table]:
+    """Extracts, processes and validates tables from a workbook based on the specified configuration.
+
+    :param workbook_data: a dictionary containing worksheet names as keys and corresponding pandas DataFrames as values
+    :param tables_config: a dictionary containing table names as keys and corresponding configuration dictionaries as
+        values
+    :return: a dictionary containing table names as keys and lists of extracted tables as pandas DataFrames as values
+    :raises ValidationError: if the data fails validation
+    """
+    extractor = ta.TableExtractor(workbook_data)
+    tables = {}
+    error_messages = []
+    for table_name, config in tables_config.items():
+        worksheet_name = config["extract"]["worksheet_name"]
+        extracted_tables = extractor.extract(**config["extract"])
+        processor = ta.TableProcessor(**config["process"])
+        validator = ta.TableValidator(config["validate"])
+        # All PFV1 tables are singular, so we assume there is only one table. This may not be true for future templates.
+        table = extracted_tables[0]
+        processor.process(table)
+        tables[table_name] = table
+        try:
+            validator.validate(table)
```

Comment:
What did you have in mind in terms of handling this error? It's occurrence would mean that the user has tampered with the template. We've typically just returned a 500 error in this case which would would show the user and error screen and pop up in Sentry for us to triage. It wouldn't be easy to summarise how to fix this issue in a message but at the same time I'm sure we could handle this kind of issue better than leaving it to the user to submit a support ticket after seeing an error screen. It may be one to hash out with the design team.

Essentially though, this is extremely unlikely to occur because the spreadsheet is protected and users typically haven't tried to change the spreadsheet like this.